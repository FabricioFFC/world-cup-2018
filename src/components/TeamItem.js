import React from 'react';
import PropTypes from 'prop-types';
import {
  StyleSheet,
  Text,
  Image,
  TouchableOpacity,
} from 'react-native';

import { withNavigation } from 'react-navigation';

const styles = StyleSheet.create({
  flag: {
    borderWidth: 1,
    borderColor: '#c7c6c6',
    flex: 1,
    height: 22,
    width: 32,
  },
  item: {
    flex: 3,
    paddingLeft: 10,
  },
  itemContainer: {
    alignItems: 'center',
    backgroundColor: '#fff',
    flex: 1,
    flexDirection: 'row',
    marginBottom: 10,
    marginRight: 10,
    paddingLeft: 10,
    paddingRight: 10,
    paddingVertical: 10,
  },
});

const TeamItem = ({ item, navigation }) => {
  const goToDetails = teamCode => navigation.navigate('TeamDetails', { teamCode });

  return (
    <TouchableOpacity onPress={() => goToDetails(item.code)} style={styles.itemContainer}>
      <Image
        style={styles.flag}
        source={{ uri: item.flag }}
      />
      <Text style={styles.item}>{item.name}</Text>
    </TouchableOpacity>
  );
};

TeamItem.propTypes = {
  item: PropTypes.shape({
    flag: PropTypes.string,
    name: PropTypes.string,
    code: PropTypes.string,
  }).isRequired,
  navigation: PropTypes.shape({
    navigate: PropTypes.func,
  }),
};

TeamItem.defaultProps = {
  navigation: undefined,
};

export default withNavigation(TeamItem);
