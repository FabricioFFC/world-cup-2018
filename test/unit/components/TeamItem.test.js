import React from 'react';
import renderer from 'react-test-renderer';
import humps from 'humps';
import { TouchableOpacity } from 'react-native';

import TeamItem from '../../../src/components/TeamItem';
import brazilMock from '../../fixtures/teams/brazil.json';

const brazilTeam = humps.camelizeKeys(brazilMock);

describe('TeamItem', () => {
  describe('render', () => {
    it('show the team item component with the team flag and name', () => {
      // given
      const team = {
        code: brazilTeam.code,
        flag: brazilTeam.flag,
        name: brazilTeam.name,
      };
      const navigateMock = jest.fn();
      const navigation = { navigate: navigateMock };
      const expectedScreenToNavigate = 'TeamDetails';
      const expectedParamsToNavigate = { teamCode: team.code };
      // when
      const teamItemRendered = renderer.create(<TeamItem item={team} navigation={navigation}/>);

      teamItemRendered.root.findByType(TouchableOpacity).props.onPress()

      const teamItemRenderedAsJSON = teamItemRendered.toJSON();
      const teamItemRenderedAsString = JSON.stringify(teamItemRenderedAsJSON);
      // then
      expect(teamItemRenderedAsString.includes(team.flag)).toBe(true);
      expect(teamItemRenderedAsString.includes(team.name)).toBe(true);
      expect(navigateMock).toBeCalledWith(expectedScreenToNavigate, expectedParamsToNavigate);
      expect(teamItemRenderedAsJSON).toMatchSnapshot();
    });

    describe('exception cases', () => {
      it('render component even when item prop is undefined', () => {
        // given
        const team = {
          code: brazilTeam.code,
          flag: brazilTeam.flag,
          name: brazilTeam.name,
        };
        const navigateMock = jest.fn();
        const navigation = { navigate: navigateMock };
        const expectedScreenToNavigate = 'TeamDetails';
        const expectedParamsToNavigate = { teamCode: undefined };
        // when
        const teamItemRendered = renderer.create(<TeamItem navigation={navigation}/>);

        teamItemRendered.root.findByType(TouchableOpacity).props.onPress()

        const teamItemRenderedAsJSON = teamItemRendered.toJSON();
        const teamItemRenderedAsString = JSON.stringify(teamItemRenderedAsJSON);
        // then
        expect(teamItemRenderedAsString.includes(team.flag)).toBe(false);
        expect(teamItemRenderedAsString.includes(team.name)).toBe(false);
        expect(navigateMock).toBeCalledWith(expectedScreenToNavigate, expectedParamsToNavigate);
        expect(teamItemRenderedAsJSON).toMatchSnapshot();
      });
    });
  });
});
