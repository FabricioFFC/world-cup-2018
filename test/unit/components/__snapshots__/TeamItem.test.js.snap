// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamItem render exception cases render component even when item prop is undefined 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "flexDirection": "row",
      "marginBottom": 10,
      "marginRight": 10,
      "opacity": 1,
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingVertical": 10,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Image
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Object {
        "borderColor": "#c7c6c6",
        "borderWidth": 1,
        "flex": 1,
        "height": 22,
        "width": 32,
      }
    }
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "flex": 3,
        "paddingLeft": 10,
      }
    }
  />
</View>
`;

exports[`TeamItem render show the team item component with the team flag and name 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "flexDirection": "row",
      "marginBottom": 10,
      "marginRight": 10,
      "opacity": 1,
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingVertical": 10,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Image
    source={
      Object {
        "uri": "http://img.fifa.com/images/flags/4/bra.png",
      }
    }
    style={
      Object {
        "borderColor": "#c7c6c6",
        "borderWidth": 1,
        "flex": 1,
        "height": 22,
        "width": 32,
      }
    }
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "flex": 3,
        "paddingLeft": 10,
      }
    }
  >
    Brazil
  </Text>
</View>
`;
